### PR TITLE
Reduce label memory

### LINKED ISSUE #14986: [Feat]: Minimize labels memory footprint
### Problem

Labels can have huge memory impact on setups with high cardinality of metrics.

Before we offload everything to disk, we should make the most to minimize the memory footprint of the existing solution.

### Description

I suggest the following changes:

## eliminate label source
Label source is not really needed. We can eliminate it.

## 1 pointer per label key:value pair
Label key:value pairs are used across many instances (charts). The ideal setup for memory optimization would be to keep just 1 pointer for every key:value pair.

To achieve this, I suggest the following:

### labels registry
We should build a global labels registry using one JudyHS.
This registry should have as key the string `key:value`, so both the key and the value provide a unique combination of items in the registry.

The structure each entry has should be:

```c
typedef struct labels_registry_entry {
   STRING *key;
   STRING *value;
   size_t refcount;
} RRDLABEL;
```

When refcount becomes zero, the entry should be removed from JudyHS and freed.

### rrdlabels structure
instead of having a DICTIONARY *rrdlabels in places we need labels, we should have this:

```c
typedef struct rrdlabels {
   SPINLOCK *spinlock;
   uint16_t used;
   uint16_t size;
   struct labels_registry_entry *array;
} RRDLABELS;
```

The spinlock should be used when adding / removing / searching entries or resizing the array.

This will give a footprint of 16 bytes overhead in places we need labels + 8 bytes per `key:value` pair.

### Importance

blocker

### Value proposition

Description|Today|Proposed|
:---:|:---:|:---:|
Overhead per RRDLABELS dictionary|336 bytes for the dictionary|16 bytes|
Overhead per unique `key:value` pair|0 bytes|24 bytes + JudyHS on `key:value`|
Memory per `key:value` pair|72 bytes + JudyHS on `key`|8 bytes|

## worst case scenario for the proposed implementation

For 1 million instances, each having 1 label with a different key:value combination:

- Today we need 336MB + 72MB = 408MB + 1 million JudyHS with 1 key each
- The proposed implementation will need: 16MB + 8MB = 24MB + 1 JudyHS having 1 million `key:value` pairs

Sharing of the same `key:value` across multiple instances will allow the proposed implementation to require a significantly smaller JudyHS too.

### Proposed implementation

See above

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
Fixes #14986

This PR reduces label memory usage (described in the linked issue)

### CODE REVIEW

File: collectors/proc.plugin/sys_block_zram.c [hunk 1]
```diff
@@ -73,7 +73,7 @@ static inline void init_rrd(const char *name, ZRAM_DEVICE *d, int update_every)
         , RRDSET_TYPE_AREA);
     d->rd_compr_data_size = rrddim_add(d->st_usage, "compressed", NULL, 1, 1024 * 1024, RRD_ALGORITHM_ABSOLUTE);
     d->rd_metadata_size = rrddim_add(d->st_usage, "metadata", NULL, 1, 1024 * 1024, RRD_ALGORITHM_ABSOLUTE);
-    rrdlabels_add(d->st_usage->rrdlabels, "device", name, RRDLABEL_SRC_AUTO);
+    rrdlabels_add(d->st_usage->rrdlabels, "device", name,RRDLABEL_SRC_AUTO);
```
Comment (thiagoftsm) [inline collectors/proc.plugin/sys_block_zram.c:76]: In normal situation we are using a  `comma` + `space`, but here I see that `space` was removed, is there a specific reason for this?
Line 111 also has this change.

Comment (stelfrag) [reply collectors/proc.plugin/sys_block_zram.c:76]: will fix

**Comment by thiagoftsm** [comment]
The PR is working as expected. Except for the warnings that should be address before to merge, the other issues are minor.

**Review by thiagoftsm — DISMISSED** [review]
After last commit, everything is working as expected, LGTM!

**Comment by MrZammler** [comment]
Hi Stelio. There are some CI fails, I think the `rrdlabels.h` file should be included in `Makefile.am` and `CMakeLists.txt` ? Not sure if it will fix them, can't reproduce the build fail localy.

**Review by MrZammler — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]

**Pull request merged**